### ISSUE TITLE
t.p.reflect.safe_repr includes the wrong traceback and misformats the return value

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3315]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#3315|
|Type|defect|
|Created|2008-06-27 13:07:51Z|
|Branch|https://github.com/twisted/twisted/tree/safe-repr-3315-2|

```
>>> from twisted.python.reflect import safe_repr
>>> class Foo(object):
...  def __repr__(self):
...   1/0
... 
>>> print safe_repr(Foo())
<Foo instance at   File "<stdin>", line 1, in <module>
  File "twisted/python/reflect.py", line 543, in safe_repr
    traceback.print_stack(file=io)
 with repr error 3082415980>
```

First, the output is obviously misformatted: it should say "<Foo instance at 3082415980 with repr error ...>". Secondly, it's using the wrong stack. Instead of print_stack()ing in the except: block, it should be print_exc()ing to get the traceback of the exception that was actually raised.

<details><summary>Searchable metadata</summary>

```
trac-id__3315 3315
type__defect defect
reporter__radix radix
priority__normal normal
milestone__ 
branch__branches_safe_repr_3315_2 branches/safe-repr-3315-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1214572071000000 1214572071000000
changetime__1227616336000000 1227616336000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [24618]) Branching to 'safe-repr-3315'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

OK, this is ready to review. I took the opportunity to make some cleanups in reflect.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

`twisted.test.test_log.LogPublisherTestCase.testSingleUnicode` fails for me with this branch, looks like for a fairly irrelevant reason:

```
===============================================================================
[FAIL]: twisted.test.test_log.LogPublisherTestCase.testSingleUnicode

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_log.py", line 254, in testSingleUnicode
    self.assertIn('with str error Traceback', self.out[0])
twisted.trial.unittest.FailTest: 'with str error Traceback' not in '2008-09-25 12:56:50-0400 [-] <unicode instance at 0x968d770 with str error:\n\t Traceback (most recent call last):\n\t  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/reflect.py", line 567, in safe_str\n\t    return str(o)\n\tUnicodeEncodeError: \'ascii\' codec can\'t encode character u\'\\xbd\' in position 7: ordinal not in range(128)\n\t>\n'
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Sorry for that, it should be better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

  1. in `twisted.python.reflect`, the changes to `AccessorType.__init__` are good and correct, I'm sure, but the code being changed has no unit tests.
  1. The case is similar for `macro`.
  1. `safe_str` has two locals, `clsName` and `className` which are redundant with each other.
  1. There's a lot of duplication between `safe_str` and `safe_repr`.  Perhaps these can both be implemented in terms of a helper function?
  1. No tests fail if I switch `unsignedID` back to `id`.
  1. in the docstring for `test_brokenClassAttribute`, can you be more specific than _broken C{__class__} attribute_?  ie, explain that it's the case where just retrieving that attribute raises an exception.  This is what the docstring says now, but I think it's a little bit subtle.
  1. likewise for `test_brokenClassNameAttribute`
  1. `test_brokenClassAttribute` could use a docstring too (not it wasn't a super awesome test to begin with, feel like adding at least some kind of `assertIn` or something?).  Actually, I suppose a lot of the tests on `SafeStr` are basically duplicates of tests on `SafeRepr`.  If the implementations of these functions get refactored into a single common helper, maybe the tests can be refactored similarly.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [24984]) Branching to 'safe-repr-3315-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

1.Tests added
 1. I didn't add tests, but deprecated macro, which doesn't seem to be used anywhere.
 1. fixed
 1. fixed
 1. fixed
 1. fixed 
 1. fixed
 1. fixed the first part, refactoring the tests looked too hard.

Back to review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

  1. I'm glad macro is deprecated now.  Shouldn't you also revert the changes to macro, though?  I hope no one is using it, but it still wouldn't be nice if we somehow broke someone's code now, particularly since now that it's deprecated the next release in the release in which we're warning them that we're _going_ to break their code. :)
  1. `_safe_format` should be `_safeFormat`
  1. in `Accessor.__init__`, the two changed lines are exercised now, but their suites aren't.  I think they should be (sorry, I hope it's easy).
  1. `test_unsignedID` could use `TestCase.patch` if you want.  Minor point, ignore if you'd rather.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

OK! I think we're good now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

There's a problem with one of the tests on 64 bit platforms:

```
===============================================================================
[FAIL]: twisted.test.test_reflect.SafeRepr.test_unsignedID

Traceback (most recent call last):
  File "/home/tarbox/projects/buildbotdir/ubuntu64-py2.5-epoll/Twisted/twisted/test/test_reflect.py", line 549, in test_unsignedID
    self.assertIn("0xffffff9c", xRepr)
twisted.trial.unittest.FailTest: '0xffffff9c' not in '<BrokenType instance at 0xffffffffffffff9c with repr error:\n Traceback (most recent call last):\n  File "/home/tarbox/projects/buildbotdir/ubuntu64-py2.5-epoll/Twisted/twisted/python/reflect.py", line 544, in _safeFormat\n    return formatter(o)\n  File "/home/tarbox/projects/buildbotdir/ubuntu64-py2.5-epoll/Twisted/twisted/test/test_reflect.py", line 462, in __repr__\n    raise RuntimeError("repr!")\nRuntimeError: repr!\n>'
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve removed owner|
|-|-|

Same player shoot again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @therve|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve set status to `closed`|
|-|-|

(In [25498]) Merge safe-repr-3315-2

Author: therve
Reviewer: exarkun
Fixes [#3315](https://github.com/twisted/twisted/issues/3315)

Fix traceback format of safe_repr and safe_str, make a bunch of cleanups in
twisted.python.reflect, deprecate macro.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
